services/nginz/third_party/nginx-zauth-module/zauth_module.c

```diff
@@ -455,11 +542,45 @@ static ngx_int_t zauth_token_typeinfo (ngx_http_request_t * r, ngx_http_variable
         }
 }
 
+static bool zauth_is_authorized_and_allowed(ngx_http_request_t * r) {
+        ZauthToken const * t = ngx_http_get_module_ctx(r, zauth_module);
+
+        if (t == NULL) {
+                return false;
+        }
+        
+        if (zauth_token_verification(t) != ZAUTH_TOKEN_VERIFICATION_SUCCESS) {
+                return false;
+        }
+
+        ZauthServerConf const * sc =
+                ngx_http_get_module_srv_conf(r, zauth_module);
+
+        if (sc == NULL || sc->acl == NULL) {
+                return false;
+        }
+
+        uint8_t is_allowed = 0;
+        
+        ngx_int_t res = zauth_token_allowed(t, sc->acl, r->uri.data, r->uri.len, &is_allowed);
+        
+        if (res != NGX_OK) {
+                return false;
+        }
+
+        return is_allowed == 1;
+}
+
 static ngx_int_t zauth_token_var (ngx_http_request_t * r, ngx_http_variable_value_t * v, uintptr_t data) {
         ZauthToken const * t = ngx_http_get_module_ctx(r, zauth_module);
         if (t == NULL) {
                 return NGX_ERROR;
         }
+        if (!zauth_is_authorized_and_allowed(r)) {
+                zauth_empty_val(v);
+                return NGX_OK;
+        }
+        
```

Comment:
call this in `zauth_handle_zauth_request` instead?

Reply:
Not exactly possible. In `zauth_handle_zauth_request` the token signature is being verified.
In `zauth_is_authorized_and_allowed` however, we check if the signature HAS BEEN verified. We do not want to do the crypto multiple times.